File: sdk/src/matlab/testDrive.m
%% Test methods in Flywheel.m
% Setup
% Before running this script, ensure the toolbox is installed
%   make sure you set SdkTestKey environment variable as user API key
%       for example: setenv('SdkTestKey', APIKEY)

% Create string to be used in testdrive
testString = '123235jakhf7sadf7v';
% A test file
filename = 'test.txt';
fid = fopen(filename, 'w');
fprintf(fid, 'This is a test file');
fclose(fid);
% Define error message
errMsg = 'Strings not equal';

% Create client
apiKey = getenv('SdkTestKey');
fw = flywheel.Flywheel(apiKey);

%% Users
disp('Testing Users')
user = fw.getCurrentUser();
assert(~isempty(user.id))

users = fw.getAllUsers();
assert(length(users) >= 1, 'No users returned')

% add a new user
email = strcat(testString, '@', testString, '.com');
userId = fw.addUser(struct('id',email,'email',email,'firstname',testString,'lastname',testString));

% modify the new user
fw.modifyUser(userId, struct('firstname', 'John'));
user2 = fw.getUser(userId);
assert(strcmp(user2.email, email), errMsg)
assert(strcmp(user2.firstname,'John'), errMsg)

fw.deleteUser(userId);

%% Groups
disp('Testing Groups')

groupId = fw.addGroup(struct('id',testString));

fw.addGroupTag(groupId, 'blue');
fw.modifyGroup(groupId, struct('label','testdrive'));

groups = fw.getAllGroups();
assert(~isempty(groups))

group = fw.getGroup(groupId);
assert(strcmp(group.tags{1},'blue'), errMsg)
assert(strcmp(group.label,'testdrive'), errMsg)

%% Projects
disp('Testing Projects')

projectId = group.addProject('label', testString);
project = fw.getProject(projectId);

project.addTag('blue');
project.update('label', 'testdrive');
project.addNote('This is a note');

projects = fw.getAllProjects();
assert(~isempty(projects), errMsg)

project.uploadFile(filename);
projectDownloadFile = fullfile(tempdir, 'download.txt');
project.downloadFile(filename, projectDownloadFile);

project = project.reload();
assert(strcmp(project.tags{1},'blue'), errMsg)
assert(strcmp(project.label,'testdrive'), errMsg)
assert(strcmp(project.notes{1}.text, 'This is a note'), errMsg)
assert(strcmp(project.files{1}.name, filename), errMsg)
s = dir(projectDownloadFile);
assert(project.files{1}.size == s.bytes, errMsg)

projectDownloadUrl = project.files{1}.url;
assert(~strcmp(projectDownloadUrl, ''), errMsg)

%% Subjects
disp('Testing Subjects')

subjectId = project.addSubject('code', testString);

fw.addSubjectTag(subjectId, 'blue');
fw.modifySubject(subjectId, struct('code', 'testdrive'));
fw.addSubjectNote(subjectId, 'This is a note');

subjects = fw.getProjectSubjects(projectId);
assert(~isempty(subjects), errMsg)

subjects = fw.getAllSubjects();
assert(~isempty(subjects), errMsg)

fw.uploadFileToSubject(subjectId, filename);
subjectDownloadFile = fullfile(tempdir, 'download2.txt');
fw.downloadFileFromSubject(subjectId, filename, subjectDownloadFile);

subject = fw.getSubject(subjectId);
assert(strcmp(subject.tags{1}, 'blue'), errMsg)
assert(strcmp(subject.label, 'testdrive'), errMsg)
assert(strcmp(subject.notes{1}.text, 'This is a note'), errMsg)
assert(strcmp(subject.files{1}.name, filename), errMsg)
s = dir(subjectDownloadFile);
assert(subject.files{1}.size == s.bytes, errMsg)

subjectDownloadUrl = fw.getSubjectDownloadUrl(subjectId, filename);
assert(~strcmp(subjectDownloadUrl, ''), errMsg)

downloadNodes = { struct('level', 'subject', 'id', subjectId) };
summary = fw.createDownloadTicket(struct('optional', false, 'nodes', downloadNodes));
assert(~isempty(summary.ticket), errMsg)
subjectDownloadTar = fullfile(tempdir, 'subject-download.tar');
fw.downloadTicket(summary.ticket, subjectDownloadTar);
s = dir(subjectDownloadTar);
assert(s.bytes >= summary.size, errMsg)

%% Sessions
disp('Testing Sessions')

sessionId = subject.addSession('label', testString);

fw.addSessionTag(sessionId, 'blue');
fw.modifySession(sessionId, struct('label', 'testdrive'));
fw.addSessionNote(sessionId, 'This is a note');

sessions = project.sessions();
assert(~isempty(sessions), errMsg)

sessions = fw.sessions();
assert(~isempty(sessions), errMsg)

fw.uploadFileToSession(sessionId, filename);
sessionDownloadFile = fullfile(tempdir, 'download2.txt');
fw.downloadFileFromSession(sessionId, filename, sessionDownloadFile);

session = fw.getSession(sessionId);
assert(strcmp(session.tags{1}, 'blue'), errMsg)
assert(strcmp(session.label, 'testdrive'), errMsg)
assert(strcmp(session.notes{1}.text, 'This is a note'), errMsg)
assert(strcmp(session.files{1}.name, filename), errMsg)
s = dir(sessionDownloadFile);
assert(session.files{1}.size == s.bytes, errMsg)

sessionDownloadUrl = fw.getSessionDownloadUrl(sessionId, filename);
assert(~strcmp(sessionDownloadUrl, ''), errMsg)

downloadNodes = { struct('level', 'session', 'id', sessionId) };
summary = fw.createDownloadTicket(struct('optional', false, 'nodes', downloadNodes));
assert(~isempty(summary.ticket), errMsg)
sessionDownloadTar = fullfile(tempdir, 'session-download.tar');
fw.downloadTicket(summary.ticket, sessionDownloadTar);
s = dir(sessionDownloadTar);
assert(s.bytes >= summary.size, errMsg)

%% Acquisitions
disp('Testing Acquisitions')

acqId = session.addAcquisition(struct('label', testString));

fw.addAcquisitionTag(acqId, 'blue');
fw.modifyAcquisition(acqId, struct('label', 'testdrive'));
fw.addAcquisitionNote(acqId, 'This is a note');

acqs = session.acquisitions();
assert(~isempty(acqs), errMsg)

acqs = fw.acquisitions();
assert(~isempty(acqs), errMsg)

fw.uploadFileToAcquisition(acqId, filename);
acquisitionDownloadFile = fullfile(tempdir, 'download3.txt');
fw.downloadFileFromAcquisition(acqId, filename, acquisitionDownloadFile);

acq = fw.getAcquisition(acqId);
assert(strcmp(acq.tags{1},'blue'), errMsg)
assert(strcmp(acq.label,'testdrive'), errMsg)
assert(strcmp(acq.notes{1}.text, 'This is a note'), errMsg)
assert(strcmp(acq.files{1}.name, filename), errMsg)
s = dir(acquisitionDownloadFile);
assert(acq.files{1}.size == s.bytes, errMsg)

acqDownloadUrl = fw.getAcquisitionDownloadUrl(acqId, filename);
assert(~strcmp(acqDownloadUrl, ''), errMsg)

acqFile = acq.files{1};

% Update file modality and type
acqFile.update('modality', 'modality', 'type', 'type');

acq = acq.reload();
acqFile = acq.files{1};
assert(strcmp(acqFile.modality, 'modality'), errMsg);
assert(strcmp(acqFile.type, 'type'), errMsg);

% Test classification functions
acqFile.replaceClassification('classification', ...
    struct('Custom', {{'measurement1', 'measurement2'}}), ...
    'modality', 'modality2');

acq = acq.reload();
acqFile = acq.files{1};
assert(strcmp(acqFile.modality, 'modality2'));
assert(strcmp(acqFile.classification.Custom{1}, 'measurement1'), errMsg);
assert(strcmp(acqFile.classification.Custom{2}, 'measurement2'), errMsg);

acqFile.updateClassification(struct('Custom', {{'HelloWorld'}}));
acqFile.deleteClassification(struct('Custom', {{'measurement2'}}));

acq = acq.reload();
acqFile = acq.files{1};
assert(strcmp(acqFile.classification.Custom{1}, 'measurement1'), errMsg);
assert(strcmp(acqFile.classification.Custom{2}, 'HelloWorld'), errMsg);

% Test file info
acqFile.replaceInfo(struct('a', 1, 'b', 2, 'c', 3, 'd', 4));
acqFile.updateInfo('c', 5);

acq = acq.reload();
acqFile = acq.files{1};
assert(acqFile.info.a == 1, errMsg);
assert(acqFile.info.b == 2, errMsg);
assert(acqFile.info.c == 5, errMsg);
assert(acqFile.info.d == 4, errMsg);

acqFile.deleteInfo({{'c', 'd'}});
acq = acq.reload();
acqFile = acq.files{1};
assert(acqFile.info.a == 1, errMsg);
assert(acqFile.info.b == 2, errMsg);
assert(~isfield(acqFile.info, 'c'), errMsg);
assert(~isfield(acqFile.info, 'd'), errMsg);

%% Collections
disp('Testing Collections')

colId = fw.addCollection(struct('label', testString));
collection = fw.getCollection(colId);

collSessions = collection.sessions();
assert(isempty(collSessions), errMsg)

collAqs = collection.acquisitions();
assert(isempty(collAqs), errMsg)

collection.addSessions(sessionId);
collection.addAcquisitions(acqId);

collSessions = collection.sessions();
assert(~isempty(collSessions), errMsg)

collAqs = collection.acquisitions();
assert(~isempty(collAqs), errMsg)

collection.addNote('This is a note');

collection.uploadFile(filename);
collectionDownloadFile = fullfile(tempdir, 'download4.txt');
collection.downloadFile(filename, collectionDownloadFile);

collection = collection.reload();
assert(strcmp(collection.notes{1}.text, 'This is a note'), errMsg)
assert(strcmp(collection.files{1}.name, filename), errMsg)
s = dir(collectionDownloadFile);
collFile = collection.files{1};
assert(collFile.size == s.bytes, errMsg)

colDownloadUrl = collFile.url;
assert(~strcmp(colDownloadUrl, ''), errMsg)

collection.deleteFile(filename);

collection = collection.reload();
assert(isempty(collection.files), errMsg)

%% Gears
disp('Testing Gears')

gearId = fw.addGear('test-drive-gear', struct('category','converter','exchange', struct('gitCommit','example','rootfsHash','sha384:example','rootfsUrl','https://example.example'),'gear', struct('name','test-drive-gear','label','Test Drive Gear','version','3','author','None','description','An empty example gear','license','Other','source','http://example.example','url','http://example.example','inputs', struct('x', struct('base','file')))));

gear = fw.getGear(gearId);
assert(strcmp(gear.gear.name, 'test-drive-gear'), errMsg)

gears = fw.getAllGears();
assert(~isempty(gears), errMsg)

job2Add = struct('gearId',gearId,'inputs',struct('x',struct('type','acquisition','id',acqId,'name',filename)));
jobId = fw.addJob(job2Add);

job = fw.getJob(jobId);
assert(strcmp(job.gearId,gearId), errMsg)

logs = fw.getJobLogs(jobId);
% Likely will not have anything in them yet

%% Containers
disp('Testing Abstract Containers')
cgroup = fw.getContainer(groupId);
assert(strcmp(group.id, cgroup.id), errMsg)

cproject = fw.getContainer(projectId);
assert(strcmp(project.id, cproject.id), errMsg)
assert(strcmp(project.label, cproject.label), errMsg)

csession = fw.getContainer(sessionId);
assert(strcmp(session.id, csession.id), errMsg)
assert(strcmp(session.label, csession.label), errMsg)

cacq = fw.getContainer(acqId);
assert(strcmp(acq.id, cacq.id), errMsg)
assert(strcmp(acq.label, cacq.label), errMsg)

fw.modifyContainer(acqId, struct('label', 'testdrive'));
cacq = fw.getContainer(acqId);
acq = fw.getAcquisition(acqId);
assert(strcmp(cacq.label,'testdrive'), errMsg)
assert(strcmp(acq.label,'testdrive'), errMsg)

fw.uploadFileToContainer(acqId, filename);
acquisitionDownloadFile = fullfile(tempdir, 'download3.txt');
fw.downloadFileFromContainer(acqId, filename, acquisitionDownloadFile);

acq = fw.getContainer(acqId);
assert(strcmp(acq.tags{1},'blue'), errMsg)
assert(strcmp(acq.label,'testdrive'), errMsg)
assert(strcmp(acq.notes{1}.text, 'This is a note'), errMsg)
assert(strcmp(acq.files{1}.name, filename), errMsg)
s = dir(acquisitionDownloadFile);
assert(acq.files{1}.size == s.bytes, errMsg)

acqDownloadUrl = fw.getContainerDownloadUrl(acqId, filename);
assert(~strcmp(acqDownloadUrl, ''), errMsg)

%% Misc
disp('Testing Misc')

config = fw.getConfig();
assert(~isempty(config), errMsg)

fwVersion = fw.getVersion();
assert(fwVersion.database >= 25, errMsg)

%% Cleanup
disp('Cleanup')

fw.deleteCollection(colId);
fw.deleteAcquisition(acqId);
fw.deleteSession(sessionId);
fw.deleteProject(projectId);
fw.deleteGroup(groupId);
fw.deleteGear(gearId);

disp('')
disp('Test drive complete.')

File: sdk/src/python/test-drive.py
#!/usr/bin/env python
from __future__ import print_function

import sys
sys.path.append('gen')

import os
import random
import string

import flywheel

from pprint import pprint

# Helper function to generate random data
def rand_string():
	return ''.join(random.choice(string.ascii_lowercase) for _ in range(10))

# Get API get from environment
api_key = os.environ['SdkTestKey']

# Create client
fw = flywheel.Flywheel(api_key)

# A test file
filepath = __file__
filename = os.path.basename(filepath)

# A second test file
filename2 = 'test.txt';
with open(filename2, 'w') as f:
    f.write('This is a test file')

#
## Users
#
user = fw.get_current_user()
assert len(user['_id']) > 0

users = fw.get_all_users()
assert len(users) >= 1

email = rand_string() + '@' + rand_string() + '.com'
userId = fw.add_user({
    'email':     email,
    'firstname': rand_string(),
    'lastname':  rand_string(),
})

fw.modify_user(userId, {'firstname': 'John'})
user2 = fw.get_user(userId)
assert user2['email'] == email
assert user2['firstname'] == 'John'

fw.delete_user(userId)


#
## Groups
#
groupId = fw.add_group({
	'_id': rand_string(),
})

fw.add_group_tag(groupId, 'blue')
fw.modify_group(groupId, {'label': 'testdrive'})

groups = fw.get_all_groups()
assert len(groups) > 0

group = fw.get_group(groupId)
assert group['tags'][0] == 'blue'
assert group['label'] == 'testdrive'


#
## Projects
#
projectId = fw.add_project({
	'label': rand_string(),
	'group': groupId,
})

fw.add_project_tag(projectId, 'blue')
fw.modify_project(projectId, {'label': 'testdrive'})
fw.add_project_note(projectId, 'This is a note')

projects = fw.get_all_projects()
assert len(projects) > 0

fw.upload_file_to_project(projectId, filepath, metadata={'zip_member_count': 15})
fw.download_file_from_project(projectId, filename, '/tmp/download.py')

project = fw.get_project(projectId)
assert project['tags'][0] == 'blue'
assert project['label'] == 'testdrive'
assert project['notes'][0]['text'] == 'This is a note'
assert project['files'][0]['name'] == filename
assert project['files'][0]['size'] == os.path.getsize('/tmp/download.py')
assert project['files'][0]['zip_member_count'] == 15

projectDownloadUrl = fw.get_project_download_url(projectId, filename)
assert projectDownloadUrl != ''

fw.delete_project_file(projectId, filename)
project = fw.get_project(projectId)
assert len(project['files']) == 0


#
## Sessions
#
sessionId = fw.add_session({
	'label': rand_string(),
	'project': projectId,
})

fw.add_session_tag(sessionId, 'blue')
fw.modify_session(sessionId, {'label': 'testdrive'})
fw.add_session_note(sessionId, 'This is a note')

sessions = fw.get_project_sessions(projectId)
assert len(sessions) > 0

sessions = fw.get_all_sessions()
assert len(sessions) > 0

fw.upload_file_to_session(sessionId, filepath, metadata={'zip_member_count': 15})
fw.download_file_from_session(sessionId, filename, '/tmp/download2.py')

session = fw.get_session(sessionId)
assert session['tags'][0] == 'blue'
assert session['label'] == 'testdrive'
assert session['notes'][0]['text'] == 'This is a note'
assert session['files'][0]['name'] == filename
assert session['files'][0]['size'] == os.path.getsize('/tmp/download2.py')
assert session['files'][0]['zip_member_count'] == 15

sessionDownloadUrl = fw.get_session_download_url(sessionId, filename)
assert sessionDownloadUrl != ''

analysisId = fw.add_session_analysis(sessionId, {
    'label': 'testdrive',
    'inputs': [{
        'type': 'session',
        'id': sessionId,
        'name': filename
    }]
})

fw.upload_output_to_analysis(analysisId, filename2)

analysis = fw.get_analysis(analysisId)
assert analysis['label'] == 'testdrive'

assert analysis['inputs'][0]['name'] == filename
assert analysis['files'][0]['name'] == filename2

fw.delete_session_file(sessionId, filename)
session = fw.get_session(sessionId)
assert len(session['files']) == 0


#
## Acquisitions
#
acqId = fw.add_acquisition({
	'label': rand_string(),
	'session': sessionId,
})

fw.add_acquisition_tag(acqId, 'blue')
fw.modify_acquisition(acqId, {'label': 'testdrive'})
fw.add_acquisition_note(acqId, 'This is a note')

acqs = fw.get_session_acquisitions(sessionId)
assert len(acqs) > 0

acqs = fw.get_all_acquisitions()
assert len(acqs) > 0

fw.upload_file_to_acquisition(acqId, filepath, metadata={'zip_member_count': 15})
fw.download_file_from_acquisition(acqId, filename, '/tmp/download3.py')

acq = fw.get_acquisition(acqId)
assert acq['tags'][0] == 'blue'
assert acq['label'] == 'testdrive'
assert acq['notes'][0]['text'] == 'This is a note'
assert acq['files'][0]['name'] == filename
assert acq['files'][0]['size'] == os.path.getsize('/tmp/download3.py')
assert acq['files'][0]['zip_member_count'] == 15

acqDownloadUrl = fw.get_acquisition_download_url(acqId, filename)
assert acqDownloadUrl != ''


#
## Collections
#
colId = fw.add_collection({
	'label': rand_string(),
	'description': rand_string(),
})

fw.add_sessions_to_collection(colId, [sessionId])
fw.add_acquisitions_to_collection(colId, [acqId])

collSessions = fw.get_collection_sessions(colId)
assert len(collSessions) == 1

collAqs = fw.get_collection_acquisitions(colId)
assert len(collAqs) == 1

fw.add_collection_note(colId, 'This is a note')

fw.upload_file_to_collection(colId, filepath, metadata={'zip_member_count': 15})
fw.download_file_from_collection(colId, filename, '/tmp/download4.py')

collection = fw.get_collection(colId)
assert collection['notes'][0]['text'] == 'This is a note'
assert collection['files'][0]['name'] == filename
assert collection['files'][0]['size'] == os.path.getsize('/tmp/download4.py')
assert collection['files'][0]['zip_member_count'] == 15

colDownloadUrl = fw.get_collection_download_url(colId, filename)
assert colDownloadUrl != ''

fw.delete_collection_file(colId, filename)
collection = fw.get_collection(colId)
assert len(collection['files']) == 0


#
## Gears
#
gearId = fw.add_gear('test-drive-gear', {
	'category': 'converter',
	'exchange' : {
		'git-commit' : 'example',
		'rootfs-hash' : 'sha384:example',
		'rootfs-url' : 'https://example.example'
	},
	'gear':	{
		'name': 'test-drive-gear',
		'label': 'Test Drive Gear',
		'version': str(random.randint(1,9000)),
		'author': 'Noone',
		'description': 'An empty example gear',
		'license': 'Other',
		'source': 'http://example.example',
		'url': 'http://example.example',
        'config': {},
		'inputs': {
			'x': {
				'base': 'file'
			}
		}
	}
})

gear = fw.get_gear(gearId)
assert gear['gear']['name'] == 'test-drive-gear'

gears = fw.get_all_gears()
assert len(gears) > 0

job_id = fw.add_job({
	'gear_id': gearId,
	'state': 'pending',
	'inputs': {
		'x': {
			'type': 'acquisition',
			'id': acqId,
			'name': 'test-drive.py'
		}
	}
})

job = fw.get_job(job_id)
assert job['gear_id'] == gearId

logs = fw.get_job_logs(job_id)
# Likely will not have anything in them yet


#
## Misc
#
config = fw.get_config()
assert config is not None

version = fw.get_version()
assert version['database'] >= 25


#
## Cleanup
#
os.remove(filename2)
fw.delete_collection(colId)
fw.delete_acquisition(acqId)
fw.delete_session(sessionId)
fw.delete_project(projectId)
fw.delete_group(groupId)
fw.delete_gear(gearId)

print('')
print('Test drive complete.')
